File: 105. Construct Binary Tree from Preorder and Inorder Traversal/105. Construct Binary Tree from Preorder and Inorder Traversal.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution(object):
    def buildTree(self, preorder, inorder):
        """
        :type preorder: List[int]
        :type inorder: List[int]
        :rtype: Optional[TreeNode]
        """
        # Create hash map for inorder values -> indices
        inMap = {}
        for i, val in enumerate(inorder):
            inMap[val] = i
            
        def buildTreeHelper(preStart, preEnd, inStart, inEnd):
            if preStart > preEnd:
                return None
                
            root = TreeNode(preorder[preStart])
            rootIndex = inMap[root.val]
            leftSubtreeSize = rootIndex - inStart
            
            root.left = buildTreeHelper(preStart + 1, 
                                      preStart + leftSubtreeSize,
                                      inStart,
                                      rootIndex - 1)
                                      
            root.right = buildTreeHelper(preStart + leftSubtreeSize + 1,
                                       preEnd, 
                                       rootIndex + 1,
                                       inEnd)
                                       
            return root
            
        return buildTreeHelper(0, len(preorder) - 1, 0, len(inorder) - 1)

File: 105. Construct Binary Tree from Preorder and Inorder Traversal/105. Construct Binary Tree from Preorder and Inorder Traversal.dart
/**
 * Definition for a binary tree node.
 * class TreeNode {
 *   int val;
 *   TreeNode? left;
 *   TreeNode? right;
 *   TreeNode([this.val = 0, this.left, this.right]);
 * }
 */
class Solution {
  late List<int> preorder;
  late List<int> inorder;
  late Map<int, int> inMap;
  
  TreeNode? buildTreeHelper(int preStart, int preEnd, int inStart, int inEnd) {
    if (preStart > preEnd) return null;
    
    TreeNode root = TreeNode(preorder[preStart]);
    int rootIndex = inMap[root.val]!;
    int leftSubtreeSize = rootIndex - inStart;
    
    root.left = buildTreeHelper(preStart + 1, preStart + leftSubtreeSize,
                               inStart, rootIndex - 1);
    root.right = buildTreeHelper(preStart + leftSubtreeSize + 1, preEnd,
                                rootIndex + 1, inEnd);
    
    return root;
  }
  
  TreeNode? buildTree(List<int> preorder, List<int> inorder) {
    this.preorder = preorder;
    this.inorder = inorder;
    this.inMap = {};
    
    for (int i = 0; i < inorder.length; i++) {
      inMap[inorder[i]] = i;
    }
    
    return buildTreeHelper(0, preorder.length - 1, 0, inorder.length - 1);
  }
}